exchanges/orderbook/incremental_updates.go

```diff
@@ -0,0 +1,234 @@
+	errDeleteFailure          = errors.New("delete update failure")
+	errInsertFailure          = errors.New("insert update failure")
+	errUpdateInsertFailure    = errors.New("update/insert update failure")
+	errRESTSnapshot           = errors.New("cannot update REST protocol loaded snapshot")
+	errChecksumMismatch       = errors.New("checksum mismatch")
+	errChecksumGeneratorUnset = errors.New("checksum generator unset")
+)
+
+// Update holds changes that are to be applied to a stored orderbook
+type Update struct {
+	UpdateID       int64
+	UpdateTime     time.Time
+	UpdatePushedAt time.Time
+	Asset          asset.Item
+	Bids           []Tranche
+	Asks           []Tranche
+	Pair           currency.Pair
+
+	// ExpectedChecksum defines the expected value when the books have been verified
+	ExpectedChecksum uint32
+	// GenerateChecksum is a function that will be called to generate a checksum from the stored orderbook post update
+	GenerateChecksum func(snapshot *Base) uint32
+	// AllowEmpty, when true, permits loading an empty order book update to set an UpdateID without including actual data
+	AllowEmpty bool
+	// Action defines the action to be performed on the orderbook e.g. amend, delete, insert, update/insert
+	// Orderbook IDs are used to identify the orderbook level to be updated, deleted or inserted
+	Action Action
+
+	SkipOutOfOrderLastUpdateID bool
+}
+
+// ProcessUpdate applies updates to the orderbook depth, on error it will invalidate the orderbook and return the
+// error, this is to ensure the orderbook is always in a valid state.
+func (d *Depth) ProcessUpdate(u *Update) error {
+	if len(u.Bids) == 0 && len(u.Asks) == 0 && !u.AllowEmpty {
+		return d.Invalidate(ErrEmptyUpdate)
+	}
+
+	// TODO: Enforce UpdatePushedAt set to determine server latency
+
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	if d.validationError != nil {
+		return d.validationError
+	}
+
+	// This will process out of order updates but will not error on them.
+	// TODO: Error on out of order updates; this is intentionally kept as is from the buffer package.
+	// Add update.UpdateTime time check to ensure that the update is newer than the last update,
+	// this should screen zero values as well.
+	if u.SkipOutOfOrderLastUpdateID && d.lastUpdateID >= u.UpdateID {
+		return nil
+	}
+
+	if d.options.restSnapshot {
+		return d.invalidate(errRESTSnapshot)
+	}
+
+	if u.Action != 0 {
+		if err := d.updateByIDAndAction(u); err != nil {
+			return d.invalidate(err)
+		}
+	} else {
+		if err := d.updateBidAskByPrice(u); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	if u.ExpectedChecksum != 0 {
+		if u.GenerateChecksum == nil {
+			return d.invalidate(errChecksumGeneratorUnset)
+		}
+		if checksum := u.GenerateChecksum(d.snapshot()); checksum != u.ExpectedChecksum {
+			return d.invalidate(fmt.Errorf("%s %s %s %w: expected '%d', got '%d'", d.exchange, d.pair, d.asset, errChecksumMismatch, u.ExpectedChecksum, checksum))
+		}
+	} else if d.verifyOrderbook {
```

Comment:
🤷 Just to check: These are mutually exclusive? We don't want to verify if we expect a checksum?

Reply:
Yeah; if we can verify the liquidity represents the exchange via the checksum then I think it should be good, `but` some exchanges only calculate checksum to ~20 length max either side, so if you have a big book it could be corrupted beyond that you wouldn't know, so there's that edge case. 

It might be safer to introduce it into library usage, push it out of the `else` and all externals have the option to turn it off anyway. 

Or optionally even have it all under a `d.verifyOrderbook` check so you can bypass checksum validation as well so you wouldn't need any of that overhead and cowboy it up. 

@thrasher- @gbjk

Reply:
Okay, so I think this is okay for now.

But to continue the discussion for another change:
Yes, I think that everything related to __validation__ [1️⃣]  should be in one place, and that checksum should probably be alongside other validation.
So that'd mean, as you suggested:
* Call a validate method always
* Conditions branch separatelly for both checksum and other verifications, so can have both, either, neither

---
1️⃣  Verfiy: Something simple truth is evident, Validate: Check something is accurate

Reply:
please re-check good sirs 🖖